### ISSUE TITLE
Annotation to control sharing mTLS certificates in pod volume

### ISSUE THREAD

**Issue by mirskifa** [istio/istio#25153]
**Describe the feature request**

If installing a service into the cluster (such as a bring-your-own Prometheus) for which envoy does not correctly identify and mTLS wrap requests, it is necessary to provide the mTLS certificates directly to the application.  A solution to this is discussed in [a recent istio.io blog post](https://istio.io/blog/2020/proxy-cert/) where a sidecar is added purely for the purpose of requesting certificates and an in-memory volume is used to communicate these to the main workload.

If the main workload also needs to be part of the mesh (for example other workloads in the cluster want to query Prometheus' datastore with mTLS enabled) then a fully fledged sidecar is required, however there is currently no way to set the `OUTPUT_CERTS` environment variable required to trigger the certificate sharing workflow.

Proposal would be to create a new sidecar annotation which allows automatically injected sidecars to utilise an in-pod volume mount to share SDS collected certificates to be shared with main workload container.

For example specifying (annotation name just a placeholder):
```yaml
sidecar.istio.io/certificateVolume: "istio-certs"
```

would result in sidecar injector adding the following to the injected sidecar container:
```yaml
env:
- name: OUTPUT_CERTS
    value: "/etc/istio-certs"
volumeMounts:
- mountPath: /etc/istio-certs
  name: istio-certs
```

This annotation would assume that the pre-injected workload already defines the volume being used (hence making the volume name part of the annotation).

**Describe alternatives you've considered**

The alternative to this injection is to manually construct a sidecar configuration (both initContainer and main sidecar proxy) that allow the pod to partake in the mesh while also configuring the `OUTPUT_CERTS` flag and volume mount to share the provisioned certificates with the main workload.  This feels much more cumbersome and brittle than an additional annotation for the sidecar injector to work with.

[x] Configuration Infrastructure
[x] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[x] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Additional context**
(This space intentionally left blank)

**Comment by JimmyCYJ** [comment]
From https://istio.io/latest/blog/2020/proxy-cert/, the certificates are shared with the service through a mounted file in tmpfs. Why do we need a in-pod volume mount instead of reusing the existing tmpfs? 
I would let @myidpt to take a look.

**Comment by mirskifa** [comment]
@JimmyCYJ the blog you linked to is the same as the one I cited in the issue.  The problem is that the sidecar template for auto-injection does not have the tmpfs mount so it would in some way need to be added.

My suggestion was to have an annotation which tells the template what volume it can use (would be added to the main workload definition before the sidecar injector hook modifies it) which is probably going to be a tmpfs location shared between the containers.  I thought it would make more sense for the definition of the volume to live in the main workload so that the workload definition would be valid whether or not the sidecar injection was enabled - the volume just wouldn’t be populated with certificates if it wasn’t.

**Comment by mlazowik** [comment]
This can be achieved by a following set of pod annotations:
```
proxy.istio.io/config: |
  proxyMetadata:
    OUTPUT_CERTS: /etc/istio-certs
sidecar.istio.io/userVolume: '[{"name": "istio-certs", "emptyDir": {"medium": "Memory"}}]'
sidecar.istio.io/userVolumeMount: '[{"name": "istio-certs", "mountPath": "/etc/istio-certs"}]'
```

Requires version `1.6.1` (`1.6.0` had a bug affecting the proxy config annotation).

**Comment by satishmane** [comment]
@mlazowik 
I have added above annotations in prometheus.yaml of prometheus operator and added certs in service monitor. It worked, prometheus is able to scrape endpoints of my microservice inside istio.

However when I checked logs next day, I saw below error in prometheus log. Do I need to add OUTPUT_CERTS env in prometheus container as well in addition to its istio side car proxy.

Scrape failed" err="Get https://<ip>:8080/actuator/prometheus: remote error: tls: expired certificate"
